src/effects/RealtimeEffectStateUI.cpp

```diff
@@ -0,0 +1,160 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  RealtimeEffectStateUI.cpp
+
+  Dmitry Vedenko
+
+**********************************************************************/
+#include "RealtimeEffectStateUI.h"
+
+#include <cassert>
+
+#include "EffectUI.h"
+#include "RealtimeEffectState.h"
+
+#include "EffectManager.h"
+#include "ProjectWindow.h"
+#include "Track.h"
+
+namespace
+{
+const RealtimeEffectState::RegisteredFactory realtimeEffectStateUIFactory { [](auto& state)
+   { return std::make_unique<RealtimeEffectStateUI>(state); }
+};
+} // namespace
+
+
+RealtimeEffectStateUI::RealtimeEffectStateUI(RealtimeEffectState& state)
+    : mRealtimeEffectState(state)
+{
+}
+
+RealtimeEffectStateUI::~RealtimeEffectStateUI()
+{
+   Hide();
+}
+
+bool RealtimeEffectStateUI::IsShown() const noexcept
+{
+   return mEffectUIHost != nullptr;
+}
+
+void RealtimeEffectStateUI::Show(AudacityProject& project)
+{
+   if (mEffectUIHost != nullptr && mEffectUIHost->IsShown())
+   {
+      // Bring the host to front
+      mEffectUIHost->Raise();
+      return;
+   }
+
+   const auto ID = mRealtimeEffectState.GetID();
+   const auto effectPlugin = EffectManager::Get().GetEffect(ID);
+
+   if (effectPlugin == nullptr)
+      return;
+
+   EffectUIClientInterface* client = effectPlugin->GetEffectUIClientInterface();
+
+   // Effect has no client interface
+   if (client == nullptr)
+      return;
+
+   // mUIHost will acquire the correct instance
+   // and will manage its lifetime
+   auto& projectWindow = ProjectWindow::Get(project);
+
+   std::shared_ptr<EffectInstance> pInstance;
+
+   auto access = mRealtimeEffectState.GetAccess();
+
+   // EffectUIHost invokes shared_from_this on access
+   Destroy_ptr<EffectUIHost> dlg(safenew EffectUIHost(
+      &projectWindow, project, *effectPlugin, *client, pInstance, *access,
+      mRealtimeEffectState.shared_from_this()));
+
+   if (!dlg->Initialize())
+      return;
+
+   // Dialog is owned by its parent now!
+   mEffectUIHost = dlg.release();
+
+   UpdateTitle();
+
+   client->ShowClientInterface(
+      projectWindow, *mEffectUIHost, mEffectUIHost->GetValidator(), false);
+
+   // The dialog was modal? That shouldn't have happened
+   if (mEffectUIHost == nullptr || !mEffectUIHost->IsShown())
+   {
+      assert(false);
+      mEffectUIHost = {};
+   }
+
+   mProjectWindowDestroyedSubscription = projectWindow.Subscribe(
+      [this](ProjectWindowDestroyedMessage) { Hide(); });
+}
+
+void RealtimeEffectStateUI::Hide()
+{
+   if (mEffectUIHost != nullptr)
+   {
+      // EffectUIHost calls Destroy in OnClose handler
+      mEffectUIHost->Close();
+      mEffectUIHost = {};
+   }
+}
+
+void RealtimeEffectStateUI::Toggle(AudacityProject& project)
+{
+   if(IsShown())
+      Hide();
+   else
+      Show(project);
+}
+
+void RealtimeEffectStateUI::UpdateTrackData(const Track& track)
```

Comment:
We come here when rebuilding the side bar.

But can there be edits of the track name that don't get reflected in the dialog title?

A complete solution might be done with subscription to undo/redo events.  Which I suggest as a bug fix after this PR, not a blocking problem.

Reply:
To be REALLY complete... it would also subcribe also to changes of preferences, in case the effect name is a built-in that can be retranslated for changed of locale.

Reply:
I had naïve expectation that TRACK_DATA_CHANGE will be fired, because, well, track data has changed. But you are right, it is not